src/lit-html.ts

```diff
@@ -12,22 +12,30 @@
  * http://polymer.github.io/PATENTS.txt
  */
 
+/**
+ * TypeScript has a problem with precompiling templates literals
+ * https://github.com/Microsoft/TypeScript/issues/17956
+ */
+const cacheTemplates = ((t: any) => t() === t())(() => ((s: TemplateStringsArray) => s)``);
+
 // The first argument to JS template tags retain identity across multiple
 // calls to a tag for the same literal, so we can cache work done per literal
 // in a Map.
-const templates = new Map<TemplateStringsArray, Template>();
+const templates = new Map<TemplateStringsArray|string, Template>();
 
 /**
  * Interprets a template literal as an HTML template that can efficiently
  * render to and update a container.
  */
-export function html(
-    strings: TemplateStringsArray, ...values: any[]): TemplateResult {
-  let template = templates.get(strings);
+export function html(strings: TemplateStringsArray, ...values: any[]): TemplateResult {
+  const key = cacheTemplates ? strings.join('{{typescriptProblems}}') : strings;
```

Comment:
This ternary is backwards. `cacheTemplates` is true when the environment has the correct behavior.